### PR TITLE
Own memory

### PR DESCRIPTION
**Pull request by ltaylor16**
One of the main features of this PR is to return std::string functions
as a Fortran ALLOCATABLE CHARACTER. This is done by passing a struct
argument to C++ to hold the address of the std::string, its data
pointer and length.  The Fortran can then ALLOCATE the string to the
correct length.  It calls a C++ helper function to copy the data.  It
also keeps track how to allocate the string.  When returning a
std::string&, it does not need to release the memory since the C++
library is responsible for it.  However, if a function returns a new
std::string (not a reference), the C wrapper allocates a temporary
std::string variable to hold the value and pass it back to Fortran.
It must then be deleted after the ALLOCATE and copying the contents.

This technique is also applied to std::vector arguments.  A temporary
std::vector is pass to C++, Fortran ALLOCATEs an array, copies the
values, then releases the std::vector argument.

Add a owner attribute to tell shroud when it should release memory.  A
deref attribute to tell if a pointer should be returned as a
type(C_PTR), ALLOCATABLE or POINTER (This is not fully implemented
yet).

### CODE REVIEW

File: docs/input.rst [hunk 1]
```diff
@@ -223,10 +199,414 @@ a brace character in the literal text, it can be escaped by doubling:
...
+              intent: in  
+    - decl: const std::string getName()
+      fattrs:
+          len: 30
+  
+
+allocatable
+^^^^^^^^^^^
+
+Sometimes it is more convient to have the wrapper allocate an
```
Comment (kennyweiss) [inline docs/input.rst]: convient -> convenient

File: docs/input.rst [hunk 2]
```diff
@@ -223,10 +199,414 @@ a brace character in the literal text, it can be escaped by doubling:
...
+``intent(out)`` array before passing it to the C++ function.  This can
+be accomplished by adding the *allocatable* attribute.  For example the
+C++ function ``cos_doubles`` takes the cosine of an ``intent(in)``
+argument and assigns it to an ``intent(out)`` argument::
+
+    void cos_doubles(double *in, double *out, int size)
+    {
+        for(int i = 0; i < size; i++) {
+            out[i] = in[i] * 2.;
```
Comment (kennyweiss) [inline docs/input.rst]: for consistency with the function name, this should probably be ``out[i] = cos(in[i])``

(Great example though!)

File: docs/input.rst [hunk 3]
```diff
@@ -223,10 +199,414 @@ a brace character in the literal text, it can be escaped by doubling:
...
+        call c_cos_doubles(in, out, sizein)
+    end subroutine cos_doubles
+
+The mold argument was added to the Fortran 2008 standard.  If the
+option **F_standard** is not 2008 then the allocate statement will be::
+
+        allocate(out(lbound(in,1):ubound(in,1)))
+
+
+For Python, same as *dimension*.
```
Comment (kennyweiss) [inline docs/input.rst]: I'm not following this sentence.  Was it meant to be more fleshed out?
(The ``dimension`` section below didn't help me, but perhaps adding a link would help?)

File: docs/installing.rst [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+.. EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+.. PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+.. PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+.. LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+.. NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+.. SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+..
+.. #######################################################################
+
+Installing
```
Comment (kennyweiss) [inline docs/installing.rst]: Thanks for adding this! It is really helpful.

File: docs/installing.rst [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+                        --cmake ${CMAKE_CURRENT_BINARY_DIR}/SetupShroud.cmake
+                        ERROR_VARIABLE SHROUD_cmake_error
+                        OUTPUT_STRIP_TRAILING_WHITESPACE )
+        if(${SHROUD_cmake_error})
+           message(FATAL_ERROR "Error from Shroud: ${SHROUD_cmake_error}")
+        endif()
+        include(${CMAKE_CURRENT_BINARY_DIR}/SetupShroud.cmake)
+    endif()
+
+The path to Shroud must defined to CMake.  It can be defined on the command line as::
```
Comment (kennyweiss) [inline docs/installing.rst]: must *be* defined?

File: docs/input.rst (hunk 2 above)
Comment (ltaylor16) [reply docs/input.rst]: This example was lifted from a web page which compared various ways to wrap.  I changed it to be a multiplication to minimize fuzz issues when comparing real numbers.  I update the test to actually use cos and fixed the comparison issues in the tests.

File: docs/input.rst (hunk 3 above)
Comment (ltaylor16) [reply docs/input.rst]: Not sure what it means either.  I added a more complete sentence but much of this still needs to be more explicit.

**Pull request merged**